client/src/main/java/io/pravega/client/stream/impl/SegmentReaderImpl.java

```diff
@@ -0,0 +1,208 @@
+        });
+    }
+
+    private ByteBuffer readEvent(long firstByteTimeoutMillis) throws EndOfSegmentException, SegmentTruncatedException, TimeoutException {
+        int read = input.read(headerReadingBuffer, firstByteTimeoutMillis);
+        if (headerReadingBuffer.hasRemaining()) {
+            log.debug("Unable to read event header for segment id {} at offset {}", input.getSegmentId(), input.getOffset());
+            log.debug("Desired event size in bytes : {}. Total bytes read : {}.", headerReadingBuffer.capacity(), read);
+            return null;
+        }
+        headerReadingBuffer.flip();
+        int type = headerReadingBuffer.getInt();
+        int length = headerReadingBuffer.getInt();
+        if (type != WireCommandType.EVENT.getCode()) {
+            throw new InvalidMessageException("Event was of wrong type: " + type);
+        }
+        if (length < 0) {
+            throw new InvalidMessageException("Event of invalid length: " + length);
+        }
+        //TODO: Here we are making the direct read call at segmentInputStream without comparing the event length with bytesInBuff.
+        // As SegmentInputStream is having buffer size as 1024 * 1024. It might be the case event size is more than buffer size.
+        ByteBuffer result = ByteBuffer.allocate(length);
+
+        readEventDataFromSegmentInputStream(result);
+        while (result.hasRemaining()) {
+            readEventDataFromSegmentInputStream(result);
+        }
+        result.flip();
+        return result;
+    }
+
+    private void readEventDataFromSegmentInputStream(ByteBuffer result) throws EndOfSegmentException, SegmentTruncatedException, TimeoutException {
+        //SSS can return empty events incase of @link{io.pravega.segmentstore.server.host.handler.PravegaRequestProcessor.ReadCancellationException}
+        if (input.read(result, PARTIAL_DATA_TIMEOUT) == 0 && result.limit() != 0) {
+            log.warn("Timeout while trying to read Event data from segment {} at offset {}. The buffer capacity is {} bytes and the data read so far is {} bytes",
+                    input.getSegmentId(), input.getOffset(), result.limit(), result.position());
+            throw new TimeoutException("Timeout while trying to read event data");
+        }
+    }
+
+    private Status getStatus(EndOfSegmentException e) {
+        Status status;
+        if (e.getErrorType().equals(EndOfSegmentException.ErrorType.END_OF_SEGMENT_REACHED)) {
+            status = Status.FINISHED;
+        } else {
+            status = Status.AVAILABLE_LATER;
+        }
+        return status;
+    }
+
+    private void handleSegmentTruncated()  {
+        log.info("{} encountered truncation for segment while read{} ", this, segment);
+        try {
+            input.setOffset(Futures.getThrowingExceptionWithTimeout(metadataClient.fetchCurrentSegmentHeadOffset(),
+                    clientConfig.getConnectTimeoutMilliSec()));
+        } catch (TimeoutException te) {
```

Comment:
Do you see there a possibility of getting other exceptions like NoSuchSegment incase of segment is sealed? are we going to handle them?
If we got Timeout exception while fetchCurrentSegmentHeadOffset that means with this logic we are not going to reset to the offset to head of stream segment and still its falls under truncated zone. In this case is the expectation is next read() call for segment reader would be throwing TruncatedDataException again to the client?